src/core/cycle_detector.go

```diff
@@ -5,108 +5,60 @@ import (
 	"strings"
 )
 
-type dependencyChain []*BuildLabel
-type dependencyLink struct {
-	from *BuildLabel
-	to   *BuildLabel
-}
-
-func (c dependencyChain) String() string {
-	labels := make([]string, len(c))
-	for i, l := range c {
-		labels[i] = l.String()
-	}
-	return strings.Join(labels, "\n -> ")
-}
-
 type cycleDetector struct {
-	deps  map[*BuildLabel][]*BuildLabel
-	queue chan dependencyLink
-}
-
-func newCycleDetector() *cycleDetector {
-	c := new(cycleDetector)
-
-	// Set this to an order of magnitude that seems sensible to avoid resizing them rapidly initially
-	c.deps = make(map[*BuildLabel][]*BuildLabel, 1000)
-
-	// Buffer this a little so handle spikes in targets coming in
-	c.queue = make(chan dependencyLink, 1000)
-
-	return c
+	graph *BuildGraph
 }
 
-// AddDependency queues up another dependency for the cycle detector to check
-func (c *cycleDetector) AddDependency(from *BuildLabel, to *BuildLabel) {
-	go func() {
-		c.queue <- dependencyLink{from: from, to: to}
-	}()
-}
-
-// checkForCycle just checks to see if there's a dependency cycle. It doesn't compute the cycle to avoid excess
-// allocations. buildCycle can be used to reconstruct the cycle once one has been found.
-func (c *cycleDetector) checkForCycle(head, tail *BuildLabel, done map[BuildLabel]struct{}) bool {
-	if _, ok := done[*tail]; ok {
-		return false
-	}
-	done[*tail] = struct{}{}
-	if tailDeps, ok := c.deps[tail]; ok {
-		for _, dep := range tailDeps {
-			if dep == head {
-				// If the tail has a dependency on the head, we've found a cycle
-				return true
-			}
-
-			if c.checkForCycle(head, dep, done) {
-				return true
+// Check runs a single check of the build graph to see if any cycles can be detected.
+// If it finds one an errCycle is returned.
+func (c *cycleDetector) Check() *errCycle {
+	log.Debug("Running cycle detection...")
+	permanent := map[*BuildTarget]struct{}{}
+	temporary := map[*BuildTarget]struct{}{}
```

Comment:
I wonder if `complete` and `partial` would make more sense as names for these. If I understand:

1. Permanent are targets where we've explored their entire dependency tree and found no cycles. They're completely visited. 
2. Temporary are targets where we're currently exploring their deps so we don't need to re-visit them. They're partially visited.

Reply:
yeah I took the names from https://en.wikipedia.org/wiki/Topological_sorting which has an example that is basically this algorithm (we're not actually top-sorting but it's pretty much the same thing).